Initial work at replacing the vitess grammar with a postgres-based one.

petermattis opened the pull request:
This is a beast, though sql/parser2 should be nearly feature complete
with sql/parser. I'm open to suggestions for how to make it more
reviewable. The code in the following files was adapted from sql/parser
with various adjustments to account for differences in the grammars:

  create.go
  delete.go
  drop.go
  encode.go
  expr.go
  insert.go
  parse.go
  rename.go
  select.go
  set.go
  show.go
  stmt.go
  truncate.go
  types.go
  union.go
  update.go
  values.go

keywords.go contains the map from keyword name to keyword ID. This was
generated using an emacs macro. I should figure out how to automatically
generate it from the grammar.

scan.go contains the new scanner (tokenizer). It was manually created
using the postgres scanner as a reference. It would have been nice if
this could have been a separate PR, but the scanner depends on the token
IDs defined in sql.go.

sql.y contains the new grammar. This is a heavily trimmed down version
of the postgres grammar. The trimming primarily occurred in removing
top-level statements (e.g. CREATE TYPE). The select statement and
expression rules are more or less identical to what postgres contains,
though the grammar does not currently contain semantic actions for
creating a nodes in the syntax tree for all rules. For example, GROUP BY
and ORDER BY are not currently handled.

See #1670.

petermattis committed:
Fork parser syntax tree.

petermattis committed:
Split up ast.go into more manageable pieces.

petermattis committed:
Add heavily trimmed down Postgres-based grammar.

petermattis committed:
Add keywords map.

petermattis committed:
Add new scanner that is compatible with the Postgres lex-based scanner.

petermattis committed:
Add semantic actions.

Flesh out types and tests.

petermattis committed:
Tweaks to the parsing tests.

petermattis committed:
Handle basic backslash escapes.

petermattis committed:
Fix make check and variable shadow warning.

spencerkimball commented on sql/parser2/encode.go:
```diff
@@ -0,0 +1,99 @@
+// Copyright 2012, Google Inc. All rights reserved.
```
What's up with this copyright notice attributed to Google, vs. others in this PR which are for Cockroach?

petermattis replied:
I really should be including both copyright notices everywhere. This code originally came from vitess, so did some of the other syntax tree code, though there are starting to be modifications.

spencerkimball replied:
Yep, makes sense.

On Wed, Jul 15, 2015 at 11:46 AM, Peter Mattis <notifications@github.com>
wrote:

> In sql/parser2/encode.go
> <https://github.com/cockroachdb/cockroach/pull/1685#discussion_r34693001>:
>
> > @@ -0,0 +1,99 @@
> > +// Copyright 2012, Google Inc. All rights reserved.
>
> I really should be including both copyright notices everywhere. This code
> originally came from vitess, so did some of the other syntax tree code,
> though there are starting to be modifications.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1685/files#r34693001>.
>

spencerkimball commented:
LGTM

petermattis committed:
Remove NullSafeEqual and change IntVal type to int64.

The postgres grammar doesn't support NullSafeEqual. We can re-add it if
desired. And postgres doesn't support unsigned types.

petermattis committed:
Update copyright notices where appropriate.

petermattis committed:
Link in log flags.

tamird commented:
Is it possible to script the creation of our grammar from the postgres grammar? I'm imagining we'll need to deal with upstream changes at some point

tamird commented on sql/parser2/update.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func (node *UpdateExpr) String() string {
+	return fmt.Sprintf("%v = %v", node.Name, node.Expr)
```
shouldn't these use `%s`? an `AndExpr` will not render correctly here, for instance

petermattis replied:
Not sure what all of the `%v` are about. Note that they come from the vitess code: https://github.com/youtube/vitess/blob/master/go/vt/sqlparser/ast.go#L958

They don't currently seem to be causing a problem, but it seems worthwhile to switch to using `%s`. Let me do so in a follow-up PR.

tamird commented:
LGTM modulo judicious use of `%v`

petermattis commented:
@tamird I don't see an easy way to script the creation of our grammar from the postgres one. We'd have to write a parser for the grammar file itself and rules for trimming it and inserting our own extensions. That sounds like a nightmare. Yes, we'll want to incorporate changes to the postgres grammar, but doing so will have to be done by hand.
